### ISSUE TITLE
Allow 'dot' in member names for representing nested relationships.

### ISSUE THREAD

**Issue by jugaadi** [json-api/json-api#1407]
`.` could be used to flatten nested objects across attributes and relationships. The server/client can easily merge the two into a single object. While the attributes maintains the object’s hierarchical structure, the relationship’s structure is flattened out by reorienting its keys to **dot** notation. This ensures compatibility with the rest of the spec in two aspects:
- Field names remain unique.
- Relationship path’s semantics remain unchanged.

 In the case of to-many relationships, the corresponding attributes can be laid out in the same order as the relationship data array. 

 For example,

```
{
  "data":{
    "type":"user",
    "id":1,
    "attributes":{
      "name":"Robert",
      "age":39,
      "school":{
        "name":"Lowell High School",
        "phone":[
          {
            "type":"office1"
          },
          {
            "type":"office2"
          },
          {
            "type":"office3"
          }
        ]
      }
    },
    "relationships":{
      "residence.address":{
        "links":{
          "self":"http://example.com/users/1/relationships/residence.address",
          "related":"http://example.com/users/1/residence.address"
        },
        "data":{
          "type":"address",
          "id":19
        }
      },
      "residence.phone":{
        "links":{
          "self":"http://example.com/users/1/relationships/residence.phone",
          "related":"http://example.com/users/1/residence.phone"
        },
        "data":[
          {
            "type":"phone",
            "id":26
          },
          {
            "type":"phone",
            "id":29
          }
        ]
      },
      "school.address":{
        "links":{
          "self":"http://example.com/users/1/relationships/school.address",
          "related":"http://example.com/users/1/school.address"
        },
        "data":{
          "type":"address",
          "id":3
        }
      },
      "school.phone":{
        "links":{
          "self":"http://example.com/users/1/relationships/school.phone",
          "related":"http://example.com/users/1/school.phone"
        },
        "data":[
          {
            "type":"phone",
            "id":11
          },
          {
            "type":"phone",
            "id":12
          },
          {
            "type":"phone",
            "id":13
          }
        ]
      }
    }
  }
}
```

PS: This doesn't handle nested arrays.

**Comment by Forbidden-Soul** [comment]
I do not see the advantage to sending data this way.

Including "residence." and "school." multiple times to child objects will increase the amount of data required to be sent by the server, rather than simply the "residence:" and "school:" objects one time as the parent object.

In your example you have increased the data size about 16 byes for 2 objects. A small difference here, but will significantly add up over a applications lifetime, or depending on the amount of objects or "dot" objects you want to send.

**Comment by jugaadi** [comment]
@Forbidden-Soul  How do you handle a relationship inside a nested attribute object? For example, in the above type, school is a nested object with some attributes and relationship fields.

**Comment by Forbidden-Soul** [comment]
@jugaadi The way I just described. Supply the parent and child objects not just simply the nested child objects. (The way you normally would).

```
"relationships": {
    "residence": {
      "address": {
        "links": {
          "self": "http://example.com/articles/1/relationships/residence.address",
          "related": "http://example.com/articles/1/residence.address"
        },
        "data": {
          "type": "address",
          "id": 19
        }
      },
      "phone": {
        "links": {
          "self": "http://example.com/articles/1/relationships/residence.phone",
          "related": "http://example.com/articles/1/residence.phone"
        },
        "data": [
          {
            "type": "phone",
            "id": 26
          },
          {
            "type": "phone",
            "id": 29
          }
        ]
      }
    },
    "school": {
      "address": {
        "links": {
          "self": "http://example.com/articles/1/relationships/school.address",
          "related": "http://example.com/articles/1/school.address"
        },
        "data": {
          "type": "address",
          "id": 3
        }
      },
      "phone": {
        "links": {
          "self": "http://example.com/articles/1/relationships/school.phone",
          "related": "http://example.com/articles/1/school.phone"
        },
        "data": [
          {
            "type": "phone",
            "id": 11
          },
          {
            "type": "phone",
            "id": 12
          },
          {
            "type": "phone",
            "id": 13
          }
        ]
      }
    }
  }
```

After receiving the object you will have no issue calling "relationships.school.phone" for example. The word "school" is only supplied once, as opposed to N times in the current implementation.
Your sudjested implementation, doesn't really offer any more ease of use, but will increase data size, that is all I am saying.

**Comment by jugaadi** [comment]
@Forbidden-Soul  

As per [spec](https://jsonapi.org/format/#document-resource-object-relationships) 
> The value of the relationships key MUST be an object (a `relationships object`)...
> A `relationship object` MUST contain at least one of the following: 
> - links : ...
> - data : ...
> - meta : ...

Your suggested implementation may not be compatible with the spec. Maybe @dgeb can confirm.

**Comment by ahx** [comment]
@jugaadi How about making "school" (look like) a real resource with something like this:

```json
{
  "data": {
    "type": "user",
    "id": "1",
    "attributes": {
      "name": "Robert",
      "age": 39
    },
    "relationships": {
      "school": {
        "links": {
          "self": "http://example.com/users/1/relationships/school",
          "related": "http://example.com/schools/19"
        },
        "data": {
          "type": "school",
          "id": "19"
        }
      },
      "address": {
        "links": {
          "self": "http://example.com/users/1/relationships/address",
          "related": "http://example.com/users/1/address"
        },
        "data": {
          "type": "address",
          "id": "19"
        }
      },
      "phones": {
        "links": {
          "self": "http://example.com/users/1/relationships/phone",
          "related": "http://example.com/users/1/phones"
        },
        "data": [
          {
            "type": "phone",
            "id": "26"
          },
          {
            "type": "phone",
            "id": "29"
          }
        ]
      }
    }
  },
  "included": [
    {
      "type": "school",
      "id": "19",
      "attributes": {
        "name": "Lowell High School"
      },
      "relationships": {
        "address": {
          "links": {
            "self": "http://example.com/schools/19/relationships/address",
            "related": "http://example.com/address/3"
          },
          "data": {
            "type": "address",
            "id": "3"
          }
        },
        "phones": {
          "links": {
            "self": "http://example.com/schools/19/relationships/phone",
            "related": "http://example.com/schools/19/phones"
          },
          "data": [
            {
              "type": "phone",
              "id": "11"
            },
            {
              "type": "phone",
              "id": "12"
            },
            {
              "type": "phone",
              "id": "13"
            }
          ]
        }
      }
    }
  ]
}
```

Note that your response can include "included" even though the client did not send an "include" parameter.

**Comment by jugaadi** [comment]
@ahx 

Since a non-resource nested object is treated as a resource object, the server MUST 
1. Either `.. return 404 Not Found when processing a request that references a related resource that does not exist.` as in the above case. 
2. Or create the nested resource as a single resource object in which case it must exist independently of the parent object and therefore, cannot be applied here.

**Comment by ahx** [comment]
@jugaadi my suggestion was that you make “schools” a (read-only) resource. The data seems to be there, so you just have to create an endpoint that returns it. 
It’s definitely more work than your initial “nested “ idea, but IMO it is easier to understand.

**Comment by jugaadi** [comment]
@ahx I wasn't aware of read only resources concept in json:api. Is it in the specs? 
Also I've noticed that you've assigned an *id* to *school* object? In our app, most nested objects do not have an *id*. The only purpose for including nested objects is to improve readability.

**Comment by ahx** [comment]
@jugaadi read only resources are definetely allowed in JSON:API. There is no extra concept there to describe this and I don't think it is needed. You just have to allow GET requests only.

> Also I've noticed that you've assigned an id to school object? 

Yes. This is needed to make [resource linkage](https://jsonapi.org/format/#document-resource-object-linkage) work in compound documents. In your case, one way to "fake" the ID could be to use the ID of the user and change the URL to something like `"http://example.com/users/1/school"`.

>  The only purpose for including nested objects is to improve readability.

I guess you are talking about human readability here. I totally understand. But if you want to optimize things for generic JSON:API clients (which there are some, but should be more of) you might consider a different strategy.

**Comment by jugaadi** [comment]
@ahx  I agree. However, this would also mean that the server has to allow creation of sub-resources(school in this case) independently of the parent resource(users in this case). This might not be a good idea considering the fact that transaction(db) semantics of the operation may get compromised.

Maybe this may get solved once bulk operations lands in the spec. 

Honestly, I still couldn't understand how relationship names can impact the performance in clients.

**Comment by jugaadi** [comment]
@cmeeren  Based on your earlier proposal(#1445 ), your feedback on this issue would be highly appreciated.

**Comment by cmeeren** [comment]
@jugaadi I may have misunderstood this issue, but I don't really have any opinions on it (other than "what's the benefit compared to the current response format with included resources").

This issue seems to be unrelated to #1445 in all respects except for the concept of using `.` as a relationship name separator. This issue seems to center around modifying the response body format using the dot notation, while #1445 only uses dots in `filter[]` parameters in order to control which resources are present in the (currently per 1.0 spec) JSON:API response.

**Comment by jugaadi** [comment]
@cmeeren Thanks for your inputs.

Currently there is no provision to express nested relationships when complex entities are transformed to JSON:API model unlike attributes(where data structures like objects/arrays are allowed). Though _included resources_ is the right way to go about it, creating, updating and deleting related resources in a single transaction is currently not possible(though atomic operation _extension_ should make it possible). 

Since lot of query params have already adopted dot operator for expressing nested objects(including your new proposal), I was wondering whether it made sense to extend it to the resource object. Current spec doesn't allow dots in field names.

PS: Im well aware that it is unrelated to your proposal. However, I presumed you came up with such a proposal to tackle issues related to complex resource models.

**Comment by cmeeren** [comment]
Am I correct to interpret this issue as relating to requests, not responses?

> I presumed you came up with such a proposal to tackle issues related to complex resource models.

Just to be clear, my motivation for that proposal was how to 1) filter top-level data based on related resource attributes (which has been discussed elsewhere), and 2) filter included resources at an arbitrary level in the same way (which I haven't seen thoroughly covered elsewhere), strengthening the querying capabilities of JSON-API (cf. GraphQL).

**Comment by jugaadi** [comment]
> Am I correct to interpret this issue as relating to requests, not responses?

Since relationship names remain the same for both req and resp, its related to both.

**Comment by cmeeren** [comment]
I'm still confused, then. Based on this:

> creating, updating and deleting related resources in a single transaction is currently not possible

I assumed this proposal was related to requests (e.g. side-loading or similar). When would you want/need to return a structure like the below in *responses*? Is there anything the below can do that can't be done by returning included resources / compound documents?

```json
"relationships":{
      "residence.address":{
        "links":{
          "self":"http://example.com/users/1/relationships/residence.address",
          "related":"http://example.com/users/1/residence.address"
        },
        "data":{
          "type":"address",
          "id":19
        }
      },
```

**Comment by jugaadi** [comment]
I faced this problem when I tried to convert an entity to its equivalent JSON:API resource. For some complex entities, the resulting resource may contain relationships fields tightly coupled with the attributes inside another attribute object. In the above example, `residence` attribute _object_ has an `address`(related resource) coupled with the rest of the attribute fields. Since relationships are supposed to be flattened, the parent object name i.e `residence` has been prefixed to the relationship name. This would work only for object data structure. For arrays, a separate resource would make more sense even though its equivalent entity doesn't exist in the server.

**Comment by cmeeren** [comment]
> I faced this problem when I tried to convert an entity to its equivalent JSON:API resource.

I don't understand what you mean by *entity* in this context.

To the best of my knowledge, in the above example, `residence` should be a separate JSON:API resource. There's no stopping you from having complex attributes, but if you have attributes that are themselves resources with relationships and all, that's exactly what stand-alone JSON:API resources are intended to model.

> For arrays, a separate resource would make more sense even though its equivalent entity doesn't exist in the server.

I'm not sure what you mean by *doesn't exist in the server*, but your JSON:API resources do not need to mirror your database, or even your domain model. JSON:API resources should be optimized for a clear and stable public API. For example, you may internally have a `customer` object with a `postNo`, `postPlace`, etc. directly on the customer object, but choose to represent it as a `customer` resource with a related `address` resource in the API.

In short, model your resources in whatever way makes the most sense for your API clients and given relevant technical back-end constraints. Legacy codebases can be hard to modernize and may require tradeoffs. But I still can't see any reason why you'd need to have a complex "object/resource attribute" that even has its own relationships, and not model it as a separate resource.

**Comment by jugaadi** [comment]
Yes. We handle it in a similar way for attribute fields with value = array objects( containing one or more relationships) but chose to use a dot notation for attribute fields with value = object( containing one or more relationships). Guess I need to redesign the models and create resources in such cases.

Thanks once again @cmeeren @ahx @Forbidden-Soul  for your comments.

**Comment by SUDDSDUDDS** [comment]
mount
